man/avahi-set-host-name.1
.TH avahi-set-host-name 1 User Manuals
.SH NAME
avahi-set-host-name \- Change mDNS host name
.SH SYNOPSIS
\fBavahi-set-host-name \fIhost-name\fB
\f1
.SH DESCRIPTION
Set the mDNS host name of a currently running Avahi daemon. The effect of this operation is not persistant across daemon restarts. This operation is usually priviliged.
.SH OPTIONS
.TP
\fB-v | --verbose\f1
Enable verbose mode.
.TP
\fB-h | --help\f1
Show help.
.TP
\fB-V | --version\f1
Show version information.
.SH AUTHORS
The avahi developers <avahi (at) lists (dot) freedesktop (dot) org>; avahi is available from \fBhttp://avahi.org/\f1
.SH SEE ALSO
\fBavahi-daemon(8)\f1
.SH COMMENTS
This man page was written using \fBxml2man(1)\f1 by Oliver Kurth.


docs/NEWS
Avahi 0.6.21
============

This is an important bugfix release.

 * Make avahi-autoipd actually produce correct ARP packets 
   (Identified and patched by Pauline Yeung)
 * Add FreeBSD kqueue support for watching /etc/avahi/services
   (Contributed by Marcus Clarke)
 * Use search domain from /etc/resolv.conf as additional browse
   domains.
 * No longer return a conflict error when two local applications
   register identical RRs.
 * Properly find alternative service/host names for very long names
 * DNS name compression fix (Sjoerd Simons)
 * Fedora init script order fix
 * Several fixes to make gcc produces less warnings
 * Minor documentation and build fixes

This release is backwards compatible with Avahi 0.6.x with x < 21.

Avahi 0.6.20
============

This is a bugfix release and contains a fix for a low risk security
vulnerability.

 * Fix a local DoS vulnerability, where an assert() could be hit by
   passing empty TXT data over D-Bus to the Avahi daemon. (Low Risk)
 * Solaris/dbm portability fixes
 * Close all open file descriptors when daemonizing. Is generally
   safer and fixes a few issues with broken init systems of several
   distributions.
 * avahi-autoipd: allow passing the path of the action script on the
   command line.
 * Several minor build fixes and other cleanups

This release is backwards compatible with Avahi 0.6.x with x < 20.

Avahi 0.6.19
============

This is a bugfix release, but also adds a new component "avahi-ui-sharp".

 * Rename zssh/zvnc to bssh/bvnc to avoid a name collision with another
   free software tool, on request of the Debian developers. I hope
   this name change early in the life of bssh/bvnc won't be too
   difficult to handle by the distributors. 
 * Add man page for bssh/bvnc
 * avahi-ui: fix segfault when browsing in empty domains
 * avahi-ui: allow GTK_RESPONSE_OK, _YES and _APPLY besides _ACCEPT as
   positive dialog response codes
 * avahi-ui-sharp: Add Mono/C# API wrapper for avahi-ui
 * Don't pick the first and the last 256 IP addresses from the IPV4LL
   range in avahi-autoipd, as required by RC3927 section 2.1
 * No longer publish the Avahi service identification cookie anymore
   by default. It was a bad idea in the first place. A better
   replacement will hopefully be made available in Avahi eventually.
 * Properly set umask before writing to /var/lib/avahi-autoipd
 * Update .deskop files to work around KDE menu issue
 * Various build fixes and other minor fixes and updates

This release is backwards compatible with Avahi 0.6.x with x < 19.

Avahi 0.6.18
============

This is a bugfix release, adds a few new features and includes a new
component.

 * Add a new library libavahi-ui which contains a standard Gtk dialog
   for browsing for services, and a new tool "zssh" that makes use of
   it, which allows browsing for SSH and VNC servers and starts ssh or
   xvncviewer if one item is selected.
 * avahi-autoipd has been ported to FreeBSD (original patch from Bruce M Simpson)
 * Improve OpenBSD, Solaris, MacOS X compatibility
 * Linux inotify support for monitoring /etc/avahi/services/ and
   /etc/avahi/hosts for changes. (Original patch from "behanw")
 * Add the ability to dump the service type database from avahi-browse
   with the new option "-b".
 * Enable GCC stack smashing protection if available and working (Patch from tedp)
 * Improve compatibility with Bonjour's libdns_sd (patch from Chris Rivera)
 * Various updates to the service type database (based on patches from uws)
 * Fix a local DoS vulnerability (very low priority, all you can do is 
   make Avahi hit an assert()); problem identified by "jamesh"
 * Fix a problem when constantly generating conflict events for an RR
 * Fix registration of very large RRs (original patch from Sjoerd Simons)
 * Various minor fixes

This release is backwards compatible with Avahi 0.6.x with x < 18.

Avahi 0.6.17
============

This is a bugfix release.

 * Don't accept "localhost" as a local mDNS host name
 * Allow running avahi-bookmarks as CGI script
 * Improve libdns_sd compatibility
 * Stability: libdns_sd mutex locking order fix
 * Publish IPv6 addresses via IPv4 and vice versa
 * IA64 fixes
 * A lot of minor cleanups and fixes

This release is backwards compatible with Avahi 0.6.x with x < 17.

Avahi 0.6.16
============

This is a bugfix release, it fixes one DoS (100% CPU Usage) and a regression
from the last release

 * Revert previous patch to check nlmsg_pid as it is bogus and breaks in
   many cases, notably when using NetworkManager (Closes: #72)
 * Replace with new SO_PASSCRED-based check of the sending UID, which
   seems to work better (Closes: #72)
 * Handle some errors in libdns_sd more gracefully the way the real
   libdns_sd does (Closes: #64)
 * Apply fix for Linux 2.6.19+ where IFA_RTA / IFLA_RTA is no longer
   defined (Closes: #86)
 * Fix doxygen comments for avahi watch, thanks to tedp (Closes: #77)
 * Make d-bus version detection work for >= 1.0 (Closes: #71)
 * Dont dbus_connection_close on shared dbus connections (Closes: #68)
 * Fix potential endless loop in dns label unpacking code (Closes: #84)
 * Fix bogus assertion in client-publish-service.c example
 * Mild fix to some doxygen docs for avahi-common/address.h
 * Fix passing in custom priviledged group (previously ignored setting)
   (Closes: #85) 

This release is backwards compatible with Avahi 0.6.x with x < 16.

Avahi 0.6.15
============

This is a bugfix release, this bug is potentially security sensitive

 * Check that netlink messages actually originate from the kernel
   and not another process.
 * Fix build on NetBSD (thanks to Daniel S. Haischt)
 * Fix dbus_service_browser not setting AVAHI_LOOKUP_RESULT_OUR_OWN.

This release is backwards compatible with Avahi 0.6.x with x < 15.

Avahi 0.6.14
============

This release fixes some bugs and includes a new component.

Changes:

  * Add new daemon "avahi-autoipd" which is an implementation of

avahi-utils/avahi-set-host-name.c
/* $Id: avahi-set-host-name.c 1512 2007-08-12 15:43:42Z lennart $ */

/***
  This file is part of avahi.
 
  avahi is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) any later version.
 
  avahi is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General
  Public License for more details.
 
  You should have received a copy of the GNU Lesser General Public
  License along with avahi; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdlib.h>
#include <stdio.h>
#include <getopt.h>
#include <assert.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <locale.h>

#include <avahi-common/simple-watch.h>
#include <avahi-common/error.h>
#include <avahi-common/malloc.h>
#include <avahi-common/domain.h>
#include <avahi-client/client.h>

#include "sigint.h"

typedef enum {
    COMMAND_UNSPEC, 
    COMMAND_HELP,
    COMMAND_VERSION
} Command;

typedef struct Config {
    int verbose;
    Command command;
} Config;

static AvahiSimplePoll *simple_poll = NULL;
static AvahiClient *client = NULL;

static void client_callback(AvahiClient *c, AvahiClientState state, AVAHI_GCC_UNUSED void * userdata) {
    switch (state) {
        case AVAHI_CLIENT_FAILURE:
            fprintf(stderr, "Client failure, exiting: %s\n", avahi_strerror(avahi_client_errno(c)));
            avahi_simple_poll_quit(simple_poll);
            break;
            
        case AVAHI_CLIENT_S_REGISTERING:
        case AVAHI_CLIENT_S_RUNNING:
        case AVAHI_CLIENT_S_COLLISION:
        case AVAHI_CLIENT_CONNECTING:
            ;
    }
}

static void help(FILE *f, const char *argv0) {
    fprintf(f,
            "%s [options] <new host name>\n\n"
            "    -h --help            Show this help\n"
            "    -V --version         Show version\n"
            "    -v --verbose         Enable verbose mode\n",
            argv0);
}

static int parse_command_line(Config *c, int argc, char *argv[]) {
    int o;

    static const struct option long_options[] = {
        { "help",           no_argument,       NULL, 'h' },
        { "version",        no_argument,       NULL, 'V' },
        { "verbose",        no_argument,       NULL, 'v' },
        { NULL, 0, NULL, 0 }
    };

    assert(c);

    c->command = COMMAND_UNSPEC;
    c->verbose = 0;

    while ((o = getopt_long(argc, argv, "hVv", long_options, NULL)) >= 0) {

        switch(o) {
            case 'h':
                c->command = COMMAND_HELP;
                break;
            case 'V':
                c->command = COMMAND_VERSION;
                break;
            case 'v':
                c->verbose = 1;
                break;
            default:
                return -1;
        }
    }

    if (c->command == COMMAND_UNSPEC) {
        if (optind != argc-1) {
            fprintf(stderr, "Invalid number of arguments, expecting exactly one.\n");
            return -1;
        }
    }
        
    return 0;
}

int main(int argc, char *argv[]) {
    int ret = 1, error;
    Config config;
    const char *argv0;

    if ((argv0 = strrchr(argv[0], '/')))
        argv0++;
    else
        argv0 = argv[0];

    if (parse_command_line(&config, argc, argv) < 0)
        goto fail;

    switch (config.command) {
        case COMMAND_HELP:
            help(stdout, argv0);
            ret = 0;
            break;
            
        case COMMAND_VERSION:
            printf("%s "PACKAGE_VERSION"\n", argv0);
            ret = 0;
            break;

        case COMMAND_UNSPEC: 
            
            if (!(simple_poll = avahi_simple_poll_new())) {
                fprintf(stderr, "Failed to create simple poll object.\n");
                goto fail;
            }
            
            if (sigint_install(simple_poll) < 0)
                goto fail;
            
            if (!(client = avahi_client_new(avahi_simple_poll_get(simple_poll), 0, client_callback, NULL, &error))) {
                fprintf(stderr, "Failed to create client object: %s\n", avahi_strerror(error));
                goto fail;
            }

            if (config.verbose) {
                const char *version, *hn;

                if (!(version = avahi_client_get_version_string(client))) {
                    fprintf(stderr, "Failed to query version string: %s\n", avahi_strerror(avahi_client_errno(client)));
                    goto fail;
                }

                if (!(hn = avahi_client_get_host_name_fqdn(client))) {
                    fprintf(stderr, "Failed to query host name: %s\n", avahi_strerror(avahi_client_errno(client)));
                    goto fail;
                }
                
                fprintf(stderr, "Server version: %s; Host name: %s\n", version, hn);
            }

            if (avahi_client_set_host_name(client, argv[optind]) < 0) {
                fprintf(stderr, "Failed to create host name resolver: %s\n", avahi_strerror(avahi_client_errno(client)));
                goto fail;
            }

            if (config.verbose) {
                const char *hn;
                
                if (!(hn = avahi_client_get_host_name_fqdn(client))) {
                    fprintf(stderr, "Failed to query host name: %s\n", avahi_strerror(avahi_client_errno(client)));
                    goto fail;
                }
                
                fprintf(stderr, "Host name successfully changed to %s\n", hn);
            }
            
            ret = 0;
            break;
    }
    
fail:

    if (client)
        avahi_client_free(client);

    sigint_uninstall();
    
    if (simple_poll)
        avahi_simple_poll_free(simple_poll);

    return ret;
}
